### PR TITLE
Permanently enabling method 3

### PR DESCRIPTION
**Pull request by igv4321**
Permanently enabling "Method 3" in HcalSimpleRecAlgo.

Some trivial cleanup of HcalSimpleRecAlgo (e.g., removed unused default constructor, added a buffer vector of doubles so that this vector does not have to be reallocated for every digi, etc).

Require proper initialization of "Method 3" in the HcalHitReconstructor module.

### CODE REVIEW

File: RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h [hunk 1]
```diff
@@ -111,8 +108,13 @@ class HcalSimpleRecAlgo {
 
   std::auto_ptr<PulseShapeFitOOTPileupCorrection> psFitOOTpuCorr_;
   
+  std::auto_ptr<PedestalSub> pedSubFxn_;
+
   // S.Brandt Feb19 : Add a pointer to the HLT algo
   std::auto_ptr<HcalDeterministicFit> hltOOTpuCorr_;
+
+  // Speed up the code a little bit by not allocating a vector every time
+  mutable std::vector<double> vectorBuffer_;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h]: What's the size of the saving from this buffer?
I'd much rather see no mutables.

Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h]: please remove this.
I don't see much benefit.

File: RecoLocalCalo/HcalRecAlgos/src/HcalSimpleRecAlgo.cc [hunk 2]
```diff
@@ -361,26 +354,26 @@ namespace HcalSimpleRecAlgoImpl {
...
+      std::vector<int> capidvec;
       for(int ip=0; ip<cs.size(); ip++){
         const int capid = digi[ip].capid();
         capidvec.push_back(capid);
       }
-      psFitOOTpuCorr->apply(cs, capidvec, calibs, correctedOutput);
-      if( correctedOutput.back() == 0 && correctedOutput.size() >1 ){
-	time = correctedOutput[1]; ampl = correctedOutput[0];
+      psFitOOTpuCorr->apply(cs, capidvec, calibs, buffer);
+      if( buffer.back() == 0 && buffer.size() >1 ){
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/HcalSimpleRecAlgo.cc]: it looks like the size is used mainly to signal the quality of the fit.
If preallocation appears to give a significant gain, just start with a fixed size vector or even an std::array and pass the fit status by value.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h (hunk 1 above)
Comment (igv4321) [inline RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h]: The saving does not come from the "size" of the vector. It is instead about not allocating (and then deallocating) the memory from the heap for each calorimeter cell processed.

File: RecoLocalCalo/HcalRecAlgos/src/HcalSimpleRecAlgo.cc (hunk 2 above)
Comment (igv4321) [inline RecoLocalCalo/HcalRecAlgos/src/HcalSimpleRecAlgo.cc]: At this point, I don't really want to change the interface of two classes written by somebody else.

File: RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h (hunk 1 above)
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h]: there are several more allocations in the methods called internally.
So, if a fix needs to be done, better do it properly.

Your change with a mutable opens a possibility to pass information between hits.
This is not good.

Cleaning and clearing is apparently done correctly now, but this makes the code much more fragile in the future developments.

Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h]: this is with 25ns hlt in 760pre5
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_7_6_0_pre5-orig.256729.step2-hlt.pp/797

i don't see any cpu cost reasons for this buffer

Comment (igv4321) [inline RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h]: The allocations internal to "Method 2" and "Method 3" classes are, of course, irrelevant to this discussion. Bug the authors of the corresponding classes if you are so inclined.

The change with a mutable, of course, does not open any possibility to pass information between hits. The vector is cleared any time it is used.

Any code is "fragile" to some degree. I believe, the advantages outweigh the drawbacks in this particular case. The "mutable" keyword is included into the language precisely for the uses similar to this.

The cost reasons are obvious: we avoid allocating the contents of the vector on the heap for every hcal rechit.

**Comment by slava77** [comment]
-1
pending code review

**Comment by fwyzard** [comment]
I do not understand lines [167](https://github.com/igv4321/cmssw/blob/always-run-hcal-method3/RecoLocalCalo/HcalRecProducers/src/HcalHitReconstructor.cc#L167) and [172](https://github.com/igv4321/cmssw/blob/always-run-hcal-method3/RecoLocalCalo/HcalRecProducers/src/HcalHitReconstructor.cc#L172) of `RecoLocalCalo/HcalRecProducers/src/HcalHitReconstructor.cc` .

**Comment by slava77** [comment]
@fwyzard 
a good question, actually,
the change was introduced back in #5389 or, originally authored by @FHead #5168 
Why was this ``` setPileupCorrection_ = 0; ``` added right after assignment ``` setPileupCorrection_ = &HcalSimpleRecAlgo::set{HF or HO}PileupCorrection; ```

**Comment by slava77** [comment]
@igv4321 
Please factor out the technical optimization (mutable vector buffer) into a separate PR.
We need the physics level feature (M3 values stored) in the release.

The optimization made is still questionable to me and it is not time critical.
Also, a more appropriate solution will not include the vector, which makes this improvement only temporary.

Thank you.

**Comment by igv4321** [comment]
Do I need to make a comment for the system to see the latest push?

**Comment by slava77** [comment]
Igor,

yes, unfortunately cms-bot doesn't notice a commit from an "old" branch and a message is needed.

Thank you very much for the factorization of the technical performance.

I noticed a change in a follow up to Andrea's comment about
 ``` setPileupCorrection_ = 0; ```
Could you please elaborate a bit on the history of the changes here.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by igv4321** [comment]
"setPileupCorrection_" variable is actually configured (as intended) by module parameters "dataOOTCorrectionName", "dataOOTCorrectionCategory", "mcOOTCorrectionName", and "mcOOTCorrectionCategory". Yi Chen, who put those lines in, simply did not realize this.

It is harmless to set this variable to 0 at those points in the code, as "Method 1" was not intended for HF and HO to begin with. While "Method 1" is currently deprecated, I don't want to completely remove this mechanism yet -- perhaps, some other method will be developed, inheriting from AbsOOTPileupCorrection.

**Comment by slava77** [comment]
@igv4321 
Do you know of the plans to add DQM/Validation plots to monitor the M3 value saved in the aux and also compare the reco (M2) and M3?

Currently, jenkins comparisons show no differences from the change made here (as expected from the lack of the DQM/validation plots)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/11507	 bb4492a
- code changes are rather minimal and look OK
- jenkins test pass and comparisons with baseline show no difference
- tested locally in CMSSW_7_6_X_2015-09-29-0900. The following numbers are from 134.805 workflow (a 25ns run 254790 from 2015C)
    - event size increases negligibly: hbhe rechit size is up by about 10%, which adds up to 0.5% in RECO, but is a factor of a few smaller in AOD (only reduced rechit collection is stored)
    - CPU time increases negligibly: hbheprereco time is up by about 2% (0.1% of all reco time)
    - the M3 energy now shows in in eaux() and appears to be correlated well with the M2 energy (::energy())
sign599_wf134p805_hbhe_energy_vs_aux_0to10

**Comment by abdoulline** [comment]
<...>
Do you know of the plans to add DQM/Validation plots to monitor the M3 value saved in the aux and also compare the reco (M2) and M3?
<...>

No, the (original) idea was to have M3 vs M2 right away without manupulation of different input files and DetId matching for on-demand comparisons (like the ones performed for "Early 2015 Physics Commissioning" task force). 
But including this kind of comparison (M3 used in HLT vs M2 used in Offline) in Offline data DQM and MC  Validation seems to be a good idea.

**Comment by davidlange6** [comment]
+1

**Pull request merged**